### PR TITLE
DeepPose alg

### PR DESCRIPTION
**Pull request by jin-s13**

### CODE REVIEW

File: mmpose/core/evaluation/top_down_eval.py [hunk 1]
```diff
@@ -393,6 +393,44 @@ def _gaussian_blur(heatmaps, kernel=11):
     return heatmaps
 
 
+def keypoints_from_regression(regression, center, scale, H, W):
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:396]: put (H, W) into and argument, such as `heatmap_size`?

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 2]
```diff
@@ -618,9 +618,11 @@ def _generate_target(self, cfg, joints_3d, joints_3d_visible):
         joint_weights = cfg['joint_weights']
         use_different_joint_weights = cfg['use_different_joint_weights']
 
-        target = joints_3d[:, :2] / image_size
-        mask = (target[:, 0] > 0) * (target[:, 0] < 1) * (target[:, 1] > 0) * (
-            target[:, 1] < 1)
+        mask = (joints_3d[:, 0] >= 0) * (
+            joints_3d[:, 0] <= image_size[0] - 1) * (joints_3d[:, 1] >= 0) * (
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:622]: turn `a <= b - 1` to `a < b`?

Comment (innerlee) [reply mmpose/datasets/pipelines/top_down_transform.py:622]: hmm unless joints_3d is a float

Comment (jin-s13) [reply mmpose/datasets/pipelines/top_down_transform.py:622]: That's float.

File: mmpose/core/evaluation/top_down_eval.py [hunk 3]
```diff
@@ -393,6 +393,42 @@ def _gaussian_blur(heatmaps, kernel=11):
...
+def keypoints_from_regression(regression, center, scale, img_size):
+    """Get final keypoint predictions from regression vectors and transform
+    them back to the image.
+
+    Note:
+        batch_size: N
+        num_keypoints: K
+
+    Args:
+        regression (np.ndarray[N, K, 2]): model prediction.
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:405]: regression -> regression_preds, or other names

File: mmpose/core/evaluation/top_down_eval.py [hunk 4]
```diff
@@ -393,6 +393,42 @@ def _gaussian_blur(heatmaps, kernel=11):
...
+
+
+    Returns:
+        preds (np.ndarray[N, K, 2]): Predicted keypoint location in images.
+        maxvals (np.ndarray[N, K, 1]): Scores (confidence) of the keypoints.
+    """
+    N, K, _ = regression.shape
+    preds, maxvals = regression, np.ones((N, K, 1), dtype=np.float32)
+
+    img_width, img_height = img_size
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:419]:
```python
img_size = list(img_size)
preds = preds * img_size  # or preds *= img_size  if you want to in-place change inputs
```

File: mmpose/core/evaluation/top_down_eval.py [hunk 5]
```diff
@@ -393,6 +393,42 @@ def _gaussian_blur(heatmaps, kernel=11):
...
+
+    preds[:, :, 0] *= img_width
+    preds[:, :, 1] *= img_height
+
+    # Transform back to the image
+    for i in range(N):
+        preds[i] = transform_preds(preds[i], center[i], scale[i],
+                                   [img_width, img_height])
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:427]: `[img_width, img_height]` -> `img_size` after above changes

File: mmpose/datasets/pipelines/top_down_transform.py [hunk 6]
```diff
@@ -591,3 +591,57 @@ def __call__(self, results):
...
+        joint_weights = cfg['joint_weights']
+        use_different_joint_weights = cfg['use_different_joint_weights']
+
+        mask = (joints_3d[:, 0] >= 0) * (
+            joints_3d[:, 0] <= image_size[0] - 1) * (joints_3d[:, 1] >= 0) * (
+                joints_3d[:, 1] <= image_size[1] - 1)
+
+        target = joints_3d[:, :2] / image_size
+
+        target = target.astype(np.float32)
```
Comment (innerlee) [inline mmpose/datasets/pipelines/top_down_transform.py:627]: if you want this line, put it before L625

File: mmpose/models/keypoint_heads/fc_head.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            img_size (tuple(img_width, img_height)): model input image size.
+        """
+        batch_size = len(img_metas)
+
+        if 'bbox_id' in img_metas[0]:
+            bbox_ids = []
+        else:
+            bbox_ids = None
+
+        c = np.zeros((batch_size, 2))
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/fc_head.py:140]: dtype

File: mmpose/models/keypoint_heads/top_down_base_head.py [hunk 8]
```diff
@@ -34,7 +36,20 @@ def forward(self, **kwargs):
...
+        Args:
+            img_metas (list(dict)): Information about data augmentation
+                By default this includes:
+                - "image_file: path to the image file
+                - "center": center of the bbox
+                - "scale": scale of the bbox
+                - "rotation": rotation of the bbox
+                - "bbox_score": score of bbox
+            output_heatmap (np.ndarray[N, K, H, W]): model predicted heatmaps.
+            img_size (tuple(img_width, img_height)): model input image size.
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/top_down_base_head.py:51]: is `img_size` used?

File: mmpose/models/losses/regression_loss.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        Args:
+            output (torch.Tensor[N, K, 2]): Output regression.
+            target (torch.Tensor[N, K, 2]): Target regression.
+            target_weight (torch.Tensor[N, K, 2]):
+                Weights across different joint types.
+        """
+        num_joints = output.size(1)
+        if self.use_target_weight:
+            loss = self.criterion(
+                output.mul(target_weight), target.mul(target_weight))
```
Comment (innerlee) [inline mmpose/models/losses/regression_loss.py:33]: What's the difference between `a*b` and `a.mul(b)`

**Comment by innerlee** [comment]
Some minor comments

File: mmpose/models/keypoint_heads/top_down_base_head.py (hunk 8 above)
Comment (jin-s13) [reply mmpose/models/keypoint_heads/top_down_base_head.py:51]: Not used for now. Just to keep the interface consistent.

File: mmpose/models/losses/regression_loss.py (hunk 9 above)
Comment (jin-s13) [reply mmpose/models/losses/regression_loss.py:33]: will change to a * b

File: mmpose/models/keypoint_heads/top_down_base_head.py (hunk 8 above)
Comment (innerlee) [reply mmpose/models/keypoint_heads/top_down_base_head.py:51]: ok

**Review by innerlee — APPROVED** [review]

**Pull request merged**